internal/services/contracts_events_consumer.go

```diff
@@ -145,3 +153,51 @@ func (c *ContractsEventsConsumer) setPrivilegeHandler(e *ContractEventData) erro
 
 	return nil
 }
+
+type MintedAftermarketDeviceArgs struct {
+	TokenID       string `mapstructure:"newTokenId"`
+	DeviceAddress string
+	Owner         string // TODO - confirm who the minter is
+}
+
+func (c *ContractsEventsConsumer) setMintedAfterMarketDevice(e *ContractEventData) error {
+	p := MintedAftermarketDeviceArgs{}
+	err := mapstructure.WeakDecode(e.Arguments, &p)
+	if err != nil {
+		return err
+	}
+
+	autopiApiService := NewAutoPiAPIService(c.settings, c.db.DBS)
+	device, err := autopiApiService.GetDeviceByEthAddress(p.DeviceAddress)
+	if err != nil {
+		c.log.Error().Msg(fmt.Sprintf("Couldn't fetch dongle with eth_address %s.", p.DeviceAddress))
+		return fmt.Errorf("couldn't fetch dongle with eth_address %s", p.DeviceAddress)
+	}
+
+	var maybeAddr null.Bytes
+
+	if strAddr := device.EthereumAddress; common.IsHexAddress(p.DeviceAddress) {
+		maybeAddr = null.BytesFrom(common.FromHex(strAddr))
+	} else {
+		c.log.Warn().Str("address", device.EthereumAddress).Msg("Invalid device Ethereum address from AutoPi.")
+	}
+
+	tokenID, _ := new(big.Int).SetString(p.TokenID, 16)
+
+	ap := models.AutopiUnit{
+		AutopiUnitID:    device.UnitID,
+		AutopiDeviceID:  null.StringFrom(device.ID),
+		EthereumAddress: maybeAddr,
+		TokenID:         types.NewNullDecimal(new(decimal.Big).SetBigMantScale(tokenID, 0)),
+	}
+
+	apCols := models.AutopiUnitColumns
+
+	err = ap.Upsert(context.Background(), c.db.DBS().Writer, true, []string{apCols.AutopiUnitID, apCols.EthereumAddress}, boil.Infer(), boil.Infer())
```

Comment:
I'm a little worried about what this will do if there's an existing record with, e.g., `user_id` populated. I never remember what it is that `boil.Infer` does.

Reply:
I guess the question then would be what fields could possibly change to require an insert. Seems the fields being passed from the contract events listener are mostly identifiers (ID's) which to me doesn't seem like something that will change easily. So do we just do an `insert` only here instead of an `upsert`

Reply:
My question is about devices that are already in our system but not minted. These days such devices are usually devices that need to generate a private key in software after getting the appropriate software update. They may have fields assigned in the database like `user_id`, but they lack an `ethereum_address` and a `token_id`.

My worry is that SQLBoiler here will blow away those fields when the minting message comes in.

Reply:
I guess the basic problem is that I don't think SQLBoiler can distinguish between "you should update this to null" and "you should leave this field alone".